rustls-cert-gen/src/main.rs

```diff
@@ -1,38 +1,109 @@
-	println!("{}", cert.serialize_private_key_pem());
-	std::fs::create_dir_all("certs/")?;
-	fs::write("certs/cert.pem", &pem_serialized.as_bytes())?;
-	fs::write("certs/cert.der", &der_serialized)?;
-	fs::write(
-		"certs/key.pem",
-		&cert.serialize_private_key_pem().as_bytes(),
-	)?;
-	fs::write("certs/key.der", &cert.serialize_private_key_der())?;
+use bpaf::Bpaf;
+use rcgen::SanType;
+use std::{net::IpAddr, path::PathBuf};
+
+mod cert;
+use cert::CertificateBuilder;
+
+fn main() -> anyhow::Result<()> {
+	let opts = options().run();
+
+	let ca = CertificateBuilder::new()
+		.signature_algorithm(&opts.sig_algo)?
+		.certificate_authority()
+		.country_name(&opts.country_name)
+		.organization_name(&opts.organization_name)
+		.build()?;
+
+	let mut entity = CertificateBuilder::new()
+		.signature_algorithm(&opts.sig_algo)?
+		.end_entity()
+		.common_name(&opts.common_name)
+		.subject_alternative_names(opts.san);
+
+	if opts.client_auth {
+		entity.client_auth();
+	};
+
+	if opts.server_auth {
+		entity.server_auth();
+	};
+
+	entity
+		.build()?
+		.serialize_pem(ca.cert())?
+		.write(&opts.output, &opts.cert_file_name)?;
+
+	ca.serialize_pem()?
+		.write(&opts.output, &opts.ca_file_name)?;
+
 	Ok(())
 }
+
+#[derive(Clone, Debug, Bpaf)]
+#[bpaf(options)]
+/// rustls-cert-gen Tls Certificate Generator
+pub struct Options {
+	/// Output Directory for generated files
+	#[bpaf(short, long, argument("output/path/"))]
+	pub output: PathBuf,
+	/// Signature algorithm
+	#[bpaf(short, long, fallback("ecdsa_p256".into()), display_fallback)]
+	pub sig_algo: String,
+	#[bpaf(external)]
+	/// Extended Key Usage Purpose: ClientAuth
+	#[bpaf(long)]
+	pub client_auth: bool,
+	/// Extended Key Usage Purpose: ServerAuth
+	#[bpaf(long)]
+	pub server_auth: bool,
+	/// Basename for end-entity cert/key
+	#[bpaf(long, fallback("cert".into()), display_fallback)]
+	pub cert_file_name: String,
+	/// Basename for ca cert/key
+	#[bpaf(long, fallback("root-ca".into()), display_fallback)]
+	pub ca_file_name: String,
+	/// Subject Alt Name (apply multiple times for multiple names/Ips)
+	#[bpaf(many, long, argument::<String>("san"), map(parse_sans))]
+	pub san: Vec<SanType>,
+	/// Common Name (Currently only used for end-entity)
+	#[bpaf(long, fallback("Tls End-Entity Certificate".into()), display_fallback)]
+	pub common_name: String,
+	/// Country Name (Currently only used for ca)
+	#[bpaf(long, fallback("BR".into()), display_fallback)]
+	pub country_name: String,
+	/// Organization Name (Currently only used for ca)
+	#[bpaf(long, fallback("Crab widgits SE".into()), display_fallback)]
+	pub organization_name: String,
+}
+
+/// Parse cli input into SanType. Try first `IpAddr`, if that fails
+/// declare it to be a DnsName.
+fn parse_sans(hosts: Vec<String>) -> Vec<SanType> {
+	hosts.into_iter().map(parse_san).collect()
+}
+
+fn parse_san(host: String) -> SanType {
+	if let Ok(ip) = host.parse::<IpAddr>() {
+		SanType::IpAddress(ip)
+	} else {
+		SanType::DnsName(host)
+	}
+}
```

Comment:
WDYT about in-lining this into `parse_sans`? I don't think it has much value on its own since only `parse_sans` uses it (excluding tests which could use `parse_sans` instead).

Reply:
done